### PR TITLE
Extend odom publisher to allow 3D

### LINKED ISSUE #934: Odometry 3D publisher plugin
## Desired behavior
I've been implementing a multicopter in ignition gazebo and I wanted to get a simple plugin that publishes the pose and velocities of the drone. I saw `odometry_publisher` existed and looked at #547. Ideally, this system would estimate the velocities in a similar way and not use the planar assumption. I want to use this plugin to obtain the ground truth of my vehicle.

## Alternatives considered
This could either be done by extending the `odometry_publisher` to handle 3D space or creating a new odometry 3d publisher. As is, the current name is misleading since its logic is derived from `gazebo_ros_planar_move` and only allows for 2D odometry.

## Implementation suggestion
Same implementation as mentioned above.

### PR DESCRIPTION
**Pull request by blakermchale**
# 🎉 New feature

Closes #934

## Summary
This adds in a new tag `<dimensions>` to the odometry publisher plugin that accepts either 2 or 3 (defaults to 2). When the value is 3 it assumes a 3D assumption instead of planar. This is useful for people who want to know the z location and angular rates. In my use case, I was using the multicopter plugin with the X3 model.

## Test it
This feature can be tested by running the `INTEGRATION_odometry_publisher` test or adding the tag to the plugin in an SDF file. I tested this manually and echoed the odometry topic while commanding the multicopter with velocities.

## Checklist
- [X] Signed all commits for DCO
- [X] Added tests
- [ ] Added example and/or tutorial
- [X] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [X] `codecheck` passed (See [contributing](https://ignitionrobotics.org/docs/all/contributing#contributing-code))
- [X] All tests passed (See [test coverage](https://ignitionrobotics.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Aignitionrobotics+repo%3Aosrf%2Fsdformat+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge**

### CODE REVIEW

**Comment by vatanaksoytezer** [comment]
I've tested this and seem to work pretty well, thanks @blakermchale

File: src/systems/odometry_publisher/OdometryPublisher.cc [hunk 1]
```diff
@@ -23,6 +23,7 @@
 #include <string>
 #include <utility>
 #include <vector>
+#include <tuple>
```
Comment (ahcorde) [inline src/systems/odometry_publisher/OdometryPublisher.cc:26]: Alphabetize

File: test/integration/odometry_publisher.cc [hunk 2]
```diff
@@ -289,6 +418,15 @@ TEST_P(OdometryPublisherTest, MovementCustomTopic)
       "/model/bar/odom");
 }
 
+/////////////////////////////////////////////////
+TEST_P(OdometryPublisherTest, Movement3d)
+{
+  TestMovement3d(
+      std::string(PROJECT_SOURCE_PATH) +
+      "/test/worlds/odometry_publisher_3d.sdf",
```
Comment (ahcorde) [inline test/integration/odometry_publisher.cc:426]: `ignition::common::JoinPath()`

File: test/integration/odometry_publisher.cc [hunk 3]
```diff
@@ -211,6 +211,135 @@ class OdometryPublisherTest : public ::testing::TestWithParam<int>
...
+    transport::Node node;
+    auto cmdVel = node.Advertise<msgs::Twist>("/X3/gazebo/command/twist");
+    node.Subscribe(_odomTopic, odomCb);
+
+    test::Relay velocityRamp;
+    math::Vector3d linVelCmd(0.5, 0.3, 1.5);
+    math::Vector3d angVelCmd(0.0, 0.0, 0.2);
+    velocityRamp.OnPreUpdate(
+        [&](const gazebo::UpdateInfo &/*_info*/,
+            gazebo::EntityComponentManager &_ecm)
```
Comment (ahcorde) [inline test/integration/odometry_publisher.cc:289]:
```suggestion
            gazebo::EntityComponentManager &/*_ecm*/)
```

**Review by ahcorde — CHANGES_REQUESTED** [review]
There are some Windows warnings. Include this change in this line https://github.com/ignitionrobotics/ign-gazebo/blob/be6c1a36142bbc0c50a2d39d397e8dd38317d83a/src/systems/odometry_publisher/OdometryPublisher.cc#L20

```
#ifdef _MSC_VER
#pragma warning(push, 0)
#endif
#include <ignition/msgs/odometry.pb.h>
#ifdef _MSC_VER
#pragma warning(pop)
#endif
```

**Comment by blakermchale** [comment]
This isn't related to the PR directly, but are there any recommendations on how to perform DCO signoffs if I missed them? I used the recommended commands and did some re-basing, however, it always ends up having conflicts and requiring me to confirm changes that I never made.

I'm fairly new to contributing to open source since I'm a college student, so any tips would be greatly appreciated.

File: src/systems/odometry_publisher/OdometryPublisher.cc [hunk 4]
```diff
@@ -138,6 +154,24 @@ void OdometryPublisher::Configure(const Entity &_entity,
     this->dataPtr->robotBaseFrame = _sdf->Get<std::string>("robot_base_frame");
   }
 
+  this->dataPtr->dimensions = 2;
+  if (!_sdf->HasElement("dimensions"))
+  {
+    ignwarn << "OdometryPublisher system plugin missing <dimensions>, "
```
Comment (chapulina) [inline src/systems/odometry_publisher/OdometryPublisher.cc:160]: Not setting dimensions is a valid use case, so we don't need to bother users with a warning. I think `igndbg` should be enough here.

**Review by chapulina — COMMENTED** [review]
Thanks for the PR, LGTM! There are just some small details to be fixed.

> are there any recommendations on how to perform DCO signoffs if I missed them? 

Be sure to sign every single commit when rebasing. You may want to squash all commits so you only need to sign one, if that makes it easier.

**Comment by chapulina** [comment]
@blakermchale , are you still interested in getting this PR in? Let us know how we can help. It would be great if you could at least sign your commits. Thanks!

**Comment by vatanaksoytezer** [comment]
@chapulina @blakermchale I'll also be more than happy to help close out this PR as currently I am using this feature and building Ignition on top of this PR.

**Comment by chapulina** [comment]
Thanks, @vatanaksoytezer , that would be great! You may need to open a new PR. We can close this one when you do.

**Comment by blakermchale** [comment]
Thanks for pinging me on this @chapulina! I'm going to close this PR and just open a new one with the changes I made signed off.

**Pull request closed without merging**